### DISCUSSION TITLE
Peeking under the covers to see next token probabilities

### DISCUSSION THREAD

**Discussion by glaforge** [tjake/Jlama#24] [Ideas]
Hi all, 

I've read this interesting [article](https://responsible-ai-developers.googleblog.com/2024/03/analyzing-next-token-probabilities-in-large-language-models.html) that shows how to use Keras and the Gemma model to peek under the hood and seen the next token probabilities.

I'm curious if 1) it's possible to do something like that when running Gemma with Jlama, and 2) if yes, how can you drive Jlama to do that?

Guillaume

**Comment by tjake** [comment]
Hi,

Yeah you can get the raw probabilities in the sample method.  

https://github.com/tjake/Jlama/blob/8d3b5ba406260402b6d3460c203afd1f2f9eb3c6/jlama-core/src/main/java/com/github/tjake/jlama/model/AbstractModel.java#L213

Currently it does argmax to pick the top value based on the temperature but it should be simple to make this method pluggable (PRs welcome!)
